import/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/CMoveService.java

```diff
@@ -0,0 +1,240 @@
+package com.google.cloud.healthcare.imaging.dicomadapter;
+
+import com.google.cloud.healthcare.IDicomWebClient;
+import com.google.cloud.healthcare.imaging.dicomadapter.cstoresender.ICStoreSender;
+import com.google.cloud.healthcare.imaging.dicomadapter.cstoresender.ICStoreSenderFactory;
+import com.google.cloud.healthcare.imaging.dicomadapter.monitoring.Event;
+import com.google.cloud.healthcare.imaging.dicomadapter.monitoring.MonitoringService;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import org.dcm4che3.data.Attributes;
+import org.dcm4che3.data.Tag;
+import org.dcm4che3.data.UID;
+import org.dcm4che3.data.VR;
+import org.dcm4che3.net.Association;
+import org.dcm4che3.net.Commands;
+import org.dcm4che3.net.Dimse;
+import org.dcm4che3.net.Status;
+import org.dcm4che3.net.pdu.PresentationContext;
+import org.dcm4che3.net.service.BasicCMoveSCP;
+import org.dcm4che3.net.service.DicomServiceException;
+import org.dcm4che3.util.TagUtils;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CMoveService extends BasicCMoveSCP {
+
+  private static Logger log = LoggerFactory.getLogger(CMoveService.class);
+  private final IDicomWebClient dicomWebClient;
+  private final AetDictionary aets;
+  private final ICStoreSenderFactory cstoreSenderFactory;
+
+  CMoveService(
+      IDicomWebClient dicomWebClient,
+      AetDictionary aets,
+      ICStoreSenderFactory cstoreSenderFactory) {
+    super(UID.StudyRootQueryRetrieveInformationModelMOVE);
+    this.dicomWebClient = dicomWebClient;
+    this.aets = aets;
+    this.cstoreSenderFactory = cstoreSenderFactory;
+  }
+
+  @Override
+  public void onDimseRQ(Association as, PresentationContext pc, Dimse dimse,
+      Attributes cmd, Attributes keys) throws IOException {
+    if (dimse != Dimse.C_MOVE_RQ) {
+      throw new DicomServiceException(Status.UnrecognizedOperation);
+    }
+
+    MonitoringService.addEvent(Event.CMOVE_REQUEST);
+
+    CMoveTask task = new CMoveTask(as, pc, cmd, keys);
+    as.getApplicationEntity().getDevice().execute(task);
+  }
+
+  private class CMoveTask extends DimseTask {
+
+    private final Attributes keys;
+
+    private CMoveTask(Association as, PresentationContext pc,
+        Attributes cmd, Attributes keys) {
+      super(as, pc, cmd);
+
+      this.keys = keys;
+    }
+
+    @Override
+    public void run() {
+      List<String> failedInstanceUids = new ArrayList<>();
```

Comment:
Considering how in http://dicom.nema.org/medical/dicom/current/output/chtml/part04/sect_C.4.2.html Failed SOP Instance UID List  is all the SOP instances for which the operation failed, (and it's even supposed to be returned for a Refused status), I think the interpretation is that this list contains instance UIDs for every instance which did not complete processing. In the code we add instances incrementally to this list if they fail processing, but I think instead we should start with a set of all SOP instance uids returned by qido, and gradually remove them as CSTOREs complete. That way in the event of a cancel or an exception, the list contains all UIDs which did not complete processing.

Reply:
I think there is important distinction between not trying a sub operation and actually failing it.

Reply:
Ok, fair enough.